### PR TITLE
lib/backend/repository: handle http and https gracefully

### PR DESCRIPTION
**Pull request by iaguis**
If docker2aci is passed the `--insecure` flag, don't directly use http but
try https first and then http.

Helps with https://github.com/coreos/rkt/issues/2036

### CODE REVIEW

**Comment by iaguis** [comment]
Tested with:

* Official Docker registry with and without `--insecure`
 * Works
* Local HTTP registry without `--insecure`:
 * `Error: conversion error: Get https://localhost:5000/v2: tls: oversized record received with length 20527`
* Local HTTP registry with `--insecure`:
 * Works
* quay with forced v1 API with and without `--insecure`:
 * Works

File: lib/backend/repository/repository.go [hunk 1]
```diff
@@ -44,18 +49,28 @@ func (rb *RepositoryBackend) GetImageInfo(url string) ([]string, *types.ParsedDo
 	dockerURL := common.ParseDockerURL(url)
 
 	var supportsV2, ok bool
+	var URLSchema string
 	if supportsV2, ok = rb.hostsV2Support[dockerURL.IndexURL]; !ok {
 		var err error
-		supportsV2, err = rb.supportsV2(dockerURL.IndexURL)
+		URLSchema, supportsV2, err = rb.supportsRegistry(dockerURL.IndexURL, "v2")
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: It isn't my turf, but I would make registry version an enum.

File: lib/backend/repository/repository.go [hunk 2]
```diff
@@ -68,10 +83,66 @@ func (rb *RepositoryBackend) BuildACI(layerNumber int, layerID string, dockerURL
...
+		url := schema + "://" + URLStr
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		rb.setBasicAuth(req)
+
+		client := &http.Client{}
+		res, err = client.Do(req)
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: Maybe just use `http.Do(req)`?

File: lib/backend/repository/repository.go [hunk 3]
```diff
@@ -68,10 +83,66 @@ func (rb *RepositoryBackend) BuildACI(layerNumber int, layerID string, dockerURL
...
+		return
+	}
+
+	schema = "https"
+	res, err := fetch(schema)
+	if err == nil {
+		ok, err = checkRegistryStatus(res.StatusCode, res.Header, version)
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: I would do `defer res.Body.Close()` here.

File: lib/backend/repository/repository.go [hunk 4]
```diff
@@ -68,10 +83,66 @@ func (rb *RepositoryBackend) BuildACI(layerNumber int, layerID string, dockerURL
...
+	}
+	if err != nil || !ok {
+		if rb.insecure {
+			if res != nil {
+				res.Body.Close()
+			}
+			schema = "http"
+			res, err = fetch(schema)
+			if err == nil {
+				ok, err = checkRegistryStatus(res.StatusCode, res.Header, version)
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: Same here.

File: lib/backend/repository/repository1.go [hunk 5]
```diff
@@ -94,7 +94,7 @@ func (rb *RepositoryBackend) buildACIV1(layerNumber int, layerID string, dockerU
 
 func (rb *RepositoryBackend) getRepoDataV1(indexURL string, remote string) (*RepoData, error) {
 	client := &http.Client{}
-	repositoryURL := rb.protocol() + path.Join(indexURL, "v1", "repositories", remote, "images")
```
Comment (krnowak) [inline lib/backend/repository/repository1.go]: Should this commit remove the `protocol` function then?

File: lib/backend/repository/repository.go (hunk 1 above)
Comment (iaguis) [inline lib/backend/repository/repository.go]: ok

File: lib/backend/repository/repository.go (hunk 2 above)
Comment (iaguis) [inline lib/backend/repository/repository.go]: there's no `http.Do()`, there's `http.Get()` but it doesn't allow messing with the headers.

File: lib/backend/repository/repository1.go (hunk 5 above)
Comment (iaguis) [inline lib/backend/repository/repository1.go]: Ooops... forgot about that

Comment (iaguis) [inline lib/backend/repository/repository1.go]: Hmm... wait, it *is* removed in this commit

**Comment by iaguis** [comment]
Updated

File: lib/backend/repository/repository.go (hunk 2 above)
Comment (krnowak) [inline lib/backend/repository/repository.go]: Ah, indeed. Nevermind then.

File: lib/backend/repository/repository1.go (hunk 5 above)
Comment (krnowak) [inline lib/backend/repository/repository1.go]: Ooops… ETOOMANYREVIEWS.

File: lib/backend/repository/repository.go [hunk 6]
```diff
@@ -14,11 +14,22 @@
...
+
 	"github.com/appc/docker2aci/lib/common"
 	"github.com/appc/docker2aci/lib/types"
 	"github.com/appc/spec/schema"
 )
 
+type registryVersion int
+
+const (
+	registryV1 registryVersion = 1 << iota
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: Just `iota` is enough, you bit-maniac. ;)

Comment (iaguis) [inline lib/backend/repository/repository.go]: But I like shifting!

File: lib/backend/repository/repository.go [hunk 7]
```diff
@@ -68,10 +91,67 @@ func (rb *RepositoryBackend) BuildACI(layerNumber int, layerID string, dockerURL
...
+
+	schema = "https"
+	res, err := fetch(schema)
+	if err == nil {
+		ok, err = checkRegistryStatus(res.StatusCode, res.Header, version)
+		defer res.Body.Close()
+	}
+	if err != nil || !ok {
+		if rb.insecure {
+			if res != nil {
```
Comment (krnowak) [inline lib/backend/repository/repository.go]: This is not needed here, since you added the defer above.

Comment (iaguis) [inline lib/backend/repository/repository.go]: oops... missed that one :disappointed:

**Comment by krnowak** [comment]
Alright, LFAD from a bystander. ;)

**Comment by iaguis** [comment]
Thanks!

**Pull request merged**